File: frontend/js/views/closing_statement.js
// Seller's closing statement generator — commission + payoff + tax proration →
// net to seller, via /calc/closing-statement. Previews live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderClosingStatement(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.cs.h1.title">// CLOSING STATEMENT</span></h1>
        <p class="muted small" data-i18n="view.cs.hint.intro">
            The seller's net sheet at a real-estate closing. From the sale price it subtracts the
            commission, the mortgage payoff, the seller's prorated share of annual property tax (days
            owed ÷ 365), and other closing costs to produce the net proceeds to the seller. Drafting aid,
            not legal/closing advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.cs.h2.inputs">Closing details</h2>
            <form id="cs-form" class="inline-form">
                <label><span data-i18n="view.cs.label.seller">Seller</span>
                    <input type="text" name="seller_name" value=""></label>
                <label><span data-i18n="view.cs.label.buyer">Buyer</span>
                    <input type="text" name="buyer_name" value=""></label>
                <label><span data-i18n="view.cs.label.property">Property address</span>
                    <input type="text" name="property_address" value=""></label>
                <label><span data-i18n="view.cs.label.price">Sale price ($)</span>
                    <input type="number" step="1000" min="0" name="sale_price_usd" value="400000" required></label>
                <label><span data-i18n="view.cs.label.closing">Closing date</span>
                    <input type="date" name="closing_date" value="2026-09-30" required></label>
                <label><span data-i18n="view.cs.label.commission">Commission (%)</span>
                    <input type="number" step="0.1" min="0" name="commission_pct" value="6"></label>
                <label><span data-i18n="view.cs.label.payoff">Mortgage payoff ($)</span>
                    <input type="number" step="100" min="0" name="mortgage_payoff_usd" value="250000"></label>
                <label><span data-i18n="view.cs.label.tax">Annual property tax ($)</span>
                    <input type="number" step="10" min="0" name="annual_property_tax_usd" value="4800"></label>
                <label><span data-i18n="view.cs.label.days">Tax days owed (of 365)</span>
                    <input type="number" step="1" min="0" max="365" name="tax_days_owed" value="180"></label>
                <label><span data-i18n="view.cs.label.other">Other closing costs ($)</span>
                    <input type="number" step="100" min="0" name="other_costs_usd" value="3000"></label>
            </form>
        </div>
        <div id="cs-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#cs-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            seller_name: (fd.get('seller_name') || '').trim(),
            buyer_name: (fd.get('buyer_name') || '').trim(),
            property_address: (fd.get('property_address') || '').trim(),
            sale_price_usd: Number(fd.get('sale_price_usd')) || 0,
            closing_date: fd.get('closing_date'),
            commission_pct: Number(fd.get('commission_pct')) || 0,
            mortgage_payoff_usd: Number(fd.get('mortgage_payoff_usd')) || 0,
            annual_property_tax_usd: Number(fd.get('annual_property_tax_usd')) || 0,
            tax_days_owed: Number(fd.get('tax_days_owed')) || 0,
            other_costs_usd: Number(fd.get('other_costs_usd')) || 0,
        };
        try {
            const doc = await api.calcClosingStatement(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.cs.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#cs-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.cs.card.net">Net to seller</div>
                    <div class="value">${money(doc.net_to_seller_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.cs.card.charges">Total charges</div>
                    <div class="value">${money(doc.total_deductions_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cs.card.proration">Tax proration</div>
                    <div class="value">${money(doc.tax_proration_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="cs-copy" type="button" data-i18n="view.cs.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="cs-download" type="button" data-i18n="view.cs.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#cs-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.cs.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.cs.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#cs-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'closing-statement.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/closing_statement.rs
//! Seller's closing / settlement statement — the net sheet at a real-estate
//! closing. From the sale price it subtracts the real-estate commission, the
//! mortgage payoff, the seller's prorated share of annual property tax (days
//! owed ÷ 365), and other closing costs to produce the net proceeds to the
//! seller. Distinct from the purchase agreement (this is the money settlement,
//! not the offer). Drafting aid, not legal/closing advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ClosingInput {
    pub seller_name: String,
    pub buyer_name: String,
    pub property_address: String,
    pub sale_price_usd: f64,
    pub closing_date: String,
    /// Real-estate commission, percent of sale price.
    #[serde(default)]
    pub commission_pct: f64,
    /// Existing mortgage payoff at closing.
    #[serde(default)]
    pub mortgage_payoff_usd: f64,
    #[serde(default)]
    pub annual_property_tax_usd: f64,
    /// Days of the tax year the seller is responsible for (prorated to closing).
    #[serde(default)]
    pub tax_days_owed: f64,
    /// Other seller closing costs (title, escrow, recording, etc.).
    #[serde(default)]
    pub other_costs_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ClosingStatement {
    pub title: String,
    pub sale_price_usd: f64,
    pub commission_usd: f64,
    pub mortgage_payoff_usd: f64,
    pub tax_proration_usd: f64,
    pub other_costs_usd: f64,
    pub total_deductions_usd: f64,
    pub net_to_seller_usd: f64,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &ClosingInput) -> ClosingStatement {
    let commission = cents(i.sale_price_usd * i.commission_pct / 100.0);
    let tax_proration = cents(i.annual_property_tax_usd * i.tax_days_owed / 365.0);
    let total_deductions =
        cents(commission + i.mortgage_payoff_usd + tax_proration + i.other_costs_usd);
    let net_to_seller = cents(i.sale_price_usd - total_deductions);

    // Seller's charges, one line each (omit zero lines except the always-shown
    // commission and payoff anchors).
    let mut lines = vec![
        format!("  Real-estate commission ({:.2}%): {}", i.commission_pct, money(commission)),
        format!("  Mortgage payoff: {}", money(i.mortgage_payoff_usd)),
    ];
    if tax_proration != 0.0 {
        lines.push(format!(
            "  Property-tax proration ({} of 365 days): {}",
            i.tax_days_owed, money(tax_proration)
        ));
    }
    if i.other_costs_usd != 0.0 {
        lines.push(format!("  Other closing costs: {}", money(i.other_costs_usd)));
    }
    lines.push(format!("Total seller charges: {}", money(total_deductions)));

    let clauses = vec![
        DocClause {
            heading: "Header".into(),
            body: format!(
                "Seller: {}\nBuyer: {}\nProperty: {}\nClosing date: {}",
                i.seller_name, i.buyer_name, i.property_address, i.closing_date
            ),
        },
        DocClause {
            heading: "1. Sale Price".into(),
            body: format!("Gross sale price: {}", money(i.sale_price_usd)),
        },
        DocClause { heading: "2. Seller's Charges".into(), body: lines.join("\n") },
        DocClause {
            heading: "3. Net Proceeds to Seller".into(),
            body: format!(
                "Net proceeds to seller (sale price {} less total charges {}): {}.",
                money(i.sale_price_usd),
                money(total_deductions),
                money(net_to_seller)
            ),
        },
        DocClause {
            heading: "Acknowledgment".into(),
            body: format!(
                "Seller: ____________________  Date: {}\n{}\n\nThis statement is for the seller's records and is not a substitute for the official settlement statement prepared by the closing agent.",
                i.closing_date, i.seller_name
            ),
        },
    ];

    ClosingStatement {
        title: "Seller's Closing Statement".into(),
        sale_price_usd: i.sale_price_usd,
        commission_usd: commission,
        mortgage_payoff_usd: i.mortgage_payoff_usd,
        tax_proration_usd: tax_proration,
        other_costs_usd: i.other_costs_usd,
        total_deductions_usd: total_deductions,
        net_to_seller_usd: net_to_seller,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> ClosingInput {
        ClosingInput {
            seller_name: "Sol Seller".into(),
            buyer_name: "Bea Buyer".into(),
            property_address: "100 Maple Ave".into(),
            sale_price_usd: 400_000.0,
            closing_date: "2026-09-30".into(),
            commission_pct: 6.0,
            mortgage_payoff_usd: 250_000.0,
            annual_property_tax_usd: 4_800.0,
            tax_days_owed: 180.0,
            other_costs_usd: 3_000.0,
        }
    }

    #[test]
    fn commission_and_proration() {
        let d = generate(&base());
        assert!(close(d.commission_usd, 24_000.0));
        assert!(close(d.tax_proration_usd, 2_367.12));
    }

    #[test]
    fn total_and_net() {
        let d = generate(&base());
        assert!(close(d.total_deductions_usd, 279_367.12));
        assert!(close(d.net_to_seller_usd, 120_632.88));
    }

    #[test]
    fn charges_clause_lists_lines() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Seller's Charges")).unwrap();
        assert!(c.body.contains("Real-estate commission (6.00%): $24000.00"));
        assert!(c.body.contains("Property-tax proration (180 of 365 days): $2367.12"));
        assert!(c.body.contains("Total seller charges: $279367.12"));
    }

    #[test]
    fn no_tax_proration_omits_line() {
        let d = generate(&ClosingInput { annual_property_tax_usd: 0.0, tax_days_owed: 0.0, ..base() });
        assert!(close(d.tax_proration_usd, 0.0));
        let c = d.clauses.iter().find(|c| c.heading.contains("Seller's Charges")).unwrap();
        assert!(!c.body.contains("proration"));
        // 24,000 + 250,000 + 3,000 = 277,000; net 123,000.
        assert!(close(d.net_to_seller_usd, 123_000.0));
    }

    #[test]
    fn net_proceeds_clause() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Net Proceeds")).unwrap();
        assert!(c.body.contains("$120632.88"));
    }

    #[test]
    fn cash_sale_no_payoff() {
        let d = generate(&ClosingInput { mortgage_payoff_usd: 0.0, ..base() });
        // 24,000 + 2,367.12 + 3,000 = 29,367.12; net 370,632.88.
        assert!(close(d.net_to_seller_usd, 370_632.88));
    }
}
